docsrc/imap/developer/overview.rst

```diff
@@ -1,52 +1,136 @@
+    Cyrus.
+
+Most ``cyd`` commands take optional switches and arguments.  For help on a
+command, you can run ``cyd help COMMAND``.
+
+The fourth command listed above, ``cyd smoke``, is a shortcut to clone, build,
+and test.  The other command you might want to run is ``cyd makedocs``, which
+will build the Cyrus documentation website, putting the built files at
+``/srv/cyrus-imapd/docsrc/build/html``.
+
+dar
+---
+
+Developing with just Docker and ``cyd`` is possible, but can be a pain.  You
+won't have your ssh credentials, so you can't push changes back up.  You'll be
+logged in a root, and you won't have your editor configuration.  To make
+development with Docker and ``cyd`` more streamlined, we've written ``dar``.
+
+``dar`` runs ``cyd`` commands inside of a container, which it manages for you.
+It expects that you'll run it from inside a git clone of the Cyrus repository.
+To get started with ``dar`` you'll want to:
+
+1.  clone the `Cyrus repository
+    <https://github.com/cyrusimap/cyrus-imapd.git>`_ and chdir into it
+
+2.  run ``dar smoke`` to build the Cyrus you've got checked out and then run
+    the tests
+
+If you try this, though, it won't work.  First, you'll be told to run ``dar
+pull``, which will pull the Docker image for your platform.  You *need* to run
+this at least once, unless you've already fetched the image by hand.  You *can*
+run it more often, to check that you're up to date.  In general, there's a new
+build of the Cyrus Docker image daily.
+
+After pulling the image, the smoke command will still fail, this time because
+you don't have a running container.  The error will tell you to run ``dar
+start``, which will start a container for this checkout.  The container will be
+running ``cyd idle``, a stub program that just keeps the container running.
+From there on, you'll use ``dar`` like you would've used ``cyd``, but from
+outside the container.  If you've been working on changes to the JMAP
+SieveScripts feature, for example, you might reach a stopping point and run::
+
+    dar build && dar test JMAPSieve
+
+Cyrus will be built using your local source (even if uncommitted) and then the
+JMAPSieve test suite will be run.  When they've finished, you'll be back at
+your local shell.  If you need to go debug a stack trace or Cassandane output,
+you can use ``dar sh`` to get a shell inside the container.
+
+When you're all done, or when you've accidentally done something to ruin or
+compromise the container, you can run ``dar prune`` to stop the container and
+remove it.  Stopping and removing the container can take a little while, if
+you've built up a lot of test output.
+
+``dar`` works by mounting your git checkout into the container where the source
+is expected.  When you build, you'll be building into that directory, meaning a
+directory on your local machine.  This means you'll send up with `foo.o` files
+in your working tree, and they'll be owned by the root user.  ``dar clean``
```

Comment:
🤔 I wonder if we can do something about this so that these files don't wind up owned by root

Reply:
I've spent a bunch of time on thinking about this so far.  There are some ideas, but nothing super promising yet. One option that I don't love is to have dar run as $UID and add that user to sudoers dynamically.